piecrust/src/call_tree.rs

```diff
@@ -0,0 +1,248 @@
+        self.0.map(|inner| unsafe {
+            let elem = (*inner).elem;
+
+            let prev = (*inner).prev;
+            if let Some(prev) = prev {
+                (*prev).children.pop();
+            }
+            free_tree(inner);
+            self.0 = prev;
+
+            elem
+        })
+    }
+
+    /// Give the current node the amount spent and recusively update amount
+    /// spent to accurately reflect what each node spent during each call.
+    pub(crate) fn update_spent(&mut self, spent: u64) {
+        if let Some(inner) = self.0 {
+            unsafe {
+                (*inner).elem.spent = spent;
+                update_spent(inner);
+            }
+        }
+    }
+
+    /// Returns the `n`th previous element from the counting from the current
+    /// node.
+    ///
+    /// The zeroth previous element is the current node.
+    pub(crate) fn nth_up(&self, n: usize) -> Option<CallTreeElem> {
+        let mut current = self.0;
+
+        let mut i = 0;
+        while i < n {
+            current = current.and_then(|inner| unsafe { (*inner).prev });
+            i += 1;
+        }
+
+        current.map(|inner| unsafe { (*inner).elem })
+    }
+
+    /// Clears the call tree of all elements.
+    pub(crate) fn clear(&mut self) {
+        unsafe {
+            if let Some(inner) = self.0 {
+                let mut root = inner;
+
+                while let Some(prev) = (*root).prev {
+                    root = prev;
+                }
+
+                self.0 = None;
+                free_tree(root);
+            }
+        }
+    }
+
+    /// Returns an iterator over the call tree, starting from the rightmost
+    /// leaf, and proceeding to the top of the current position of the tree.
+    pub fn iter(&self) -> impl Iterator<Item = &CallTreeElem> {
+        CallTreeIter {
+            tree: self.0.map(|root| unsafe {
+                let mut node = root;
+
+                while !(*node).children.is_empty() {
+                    let last_index = (*node).children.len() - 1;
+                    node = (*node).children[last_index];
+                }
+
+                Subtree { root, node }
+            }),
+            _marker: PhantomData,
+        }
+    }
+}
+
+struct Subtree {
+    root: *mut CallTreeNode,
+    node: *mut CallTreeNode,
+}
+
+impl Drop for CallTree {
+    fn drop(&mut self) {
+        self.clear()
+    }
+}
+
+unsafe fn update_spent(node: *mut CallTreeNode) {
+    let node = &mut *node;
+    node.children.iter_mut().for_each(|&mut child| unsafe {
+        node.elem.spent -= (*child).elem.spent;
+        update_spent(child);
+    });
+}
+
+unsafe fn free_tree(root: *mut CallTreeNode) {
+    let mut node = Box::from_raw(root);
+
+    let mut children = Vec::new();
+    mem::swap(&mut node.children, &mut children);
+
+    for child in children {
+        free_tree(child);
+    }
+}
+
+struct CallTreeNode {
+    elem: CallTreeElem,
+    children: Vec<*mut Self>,
+    prev: Option<*mut Self>,
```

Comment:
Wouldn't it make sense to say `parent` instead of `prev`?

Reply:
Agreed. Substituted for `parent`, including all variables and mentions in comments.